### PR TITLE
Draft: Joram scrapper code base

### LINKED ISSUE #5: Retrieve gas prices
## Expected Behavior

The prices should be retrieved from the [JORAM](https://joram.madeira.gov.pt/joram/2serie/) website. The weekly gas prices are issued every Friday, in **2.º Suplemento**. You can check some examples here:

image

## Actual Behavior

Right now, the gas prices are retrieved from the [Direção Regional de Economia e Transportes Terrestres](https://www.madeira.gov.pt/drett) website. However, these prices are not updated every Friday (as the JORAM document is issued with gas prices for the next week).

To check how we retrieve the gas prices at this moment, check [this file](https://github.com/carlosrsabreu/devo-abastecer/blob/main/script.py
).

### PR DESCRIPTION
**Pull request by 13dev**
### What this PR does

1. Adds a simple base to search gas prices in the joram website. This is possible by doing a conversion (pdf -> text) and match regex (lowest steps possible) line by line.

Note the first TODO there are some ideas to finish the implementation

### Where should the reviewer start

1. This only adds a first step to finish #5 

### Related issues

Related #5

### CODE REVIEW

**Review by joaoofreitas — COMMENTED** [review]
Nice work @13dev! Simple yet so functional.

I would just add a regex for relating the prices to the product that has been sold. 
Maybe store and return a dictionary?

Then we would have a complete _joram_ parser module that we just need to import.

What do you guys think?
@13dev @carlosrsabreu @Dntfreitas

**Comment by Dntfreitas** [comment]
Yes, it also simplify the integration with the current API

**Comment by 13dev** [comment]
Thanks 4 the feedback.

I've done some modifications:

- Every price is yield
- Change dir of scrapper to root (so we can use constants.py)
- Add a function to associate prices with gas names

Running the script you will get something like this:

`{'Gasolina IO95': '1,889', 'Gasóleo Rodoviário': '1,789', 'Gasóleo Colorido e Marcado': '1,425'}`

**Comment by 13dev** [comment]
Update: Perfomance improvement

The script will stop searching for a price when we already have the 3 prices

**Review by joaoofreitas — CHANGES_REQUESTED** [review]
In general looks good to me. 

I'm sorry to bother about this, but how can we guarantee that the labels correspond to the price?

I suggest adding some verification in `regex` when adding the labels.
Then we can guarantee that the price parsed can indeed correspond to the label.

_We never know when JORAM can change the document format._

**Comment by joaoofreitas** [comment]
If you need any help or want do discuss it further, just contact me. :)

Thanks to your help! @13dev
Legend.

**Comment by 13dev** [comment]
Good point!

Changed the regex to match with the label, much cleaner code!

**Comment by joaoofreitas** [comment]
What a _demon_ of _regex_ right there! 

Needed to test it myself. Works flawlessly.

From my side, it's an approval. Someone need to integrate his changes into the codebase afterwards as a module. 
_(@carlosrsabreu please create an issue if you want to assign it to me)_

Great work, shall be merged soon enough. @13dev 
Now we wait until Thursday, so @carlosrsabreu can approve the changes.

Any thoughts? @carlosrsabreu @Dntfreitas @13dev

File: constants.py
Comment (13dev) [reply constants.py:31]: i agree, surely have better way to do!

2º group can be `(?:.+)` but it adds 100 more steps

**Comment by carlosrsabreu** [comment]
@13dev,

Thank you for this PR; it's working perfectly! It's **so great** to have local developers interested in developing open-source community projects 😁

Thank you, @joaoofreitas and @Dntfreitas, for your thoughts and comments!

I will create an issue to integrate this scrapper into the codebase with the mentioned TODOs.

**Review by carlosrsabreu — APPROVED** [review]

**Pull request merged**
